keras_cv/models/segmentation/segment_anything/sam_prompt_encoder.py

```diff
@@ -0,0 +1,321 @@
+        )
+        self.no_mask_embed = keras.layers.Embedding(1, embed_dim)
+
+        for layer in [
+            self.foreground_point_embed,
+            self.background_point_embed,
+            self.top_left_corner_embed,
+            self.bottom_right_corner_embed,
+            self.not_a_point_embed,
+            self.no_mask_embed,
+        ]:
+            layer.build(None)
+
+        self.built = True
+
+    def get_dense_pe(self):
+        """Get positional embeddings for the image being prompted.
+
+        Returns:
+            tensor: The positional embeddings of the image.
+        """
+        # convert the image_embedding_size to a tensor since keras core
+        # expects the input type to be a symbolic/concrete tensor.
+        return self.positional_embedding_layer(
+            ops.convert_to_tensor(self.image_embedding_size, dtype="float32")
+        )[None, ...]
+
+    def __embed_points(self, points, labels, pad):
+        points = points + 0.5
+        if pad:
+            padding_point = ops.zeros((points.shape[0], 1, 2), dtype=self.dtype)
+            padding_label = -ops.ones((labels.shape[0], 1), dtype=self.dtype)
+            points = ops.concatenate([points, padding_point], axis=1)
+            labels = ops.concatenate([labels, padding_label], axis=1)
+        point_embeddings = self.positional_embedding_layer.call_with_coords(
+            points, self.input_image_size
+        )
+        labels = ops.broadcast_to(labels[..., None], point_embeddings.shape)
+        point_embeddings = ops.where(
+            labels == 0,
+            point_embeddings + self.background_point_embed.weights[0],
+            point_embeddings + self.foreground_point_embed.weights[0],
+        )
+        point_embeddings = ops.where(
+            labels == -1,
+            # TODO: for whatever reason, ops.broadcast_to doesn't work here, so
+            #       we instead use zeros_like to broadcast to the correct shape.
+            self.not_a_point_embed.weights[0]
+            + ops.zeros_like(point_embeddings),
+            point_embeddings,
+        )
+        return point_embeddings
+
+    def __embed_box(self, box):
+        box = box + 0.5
+        coords = ops.reshape(box, (-1, 2, 2))
+        corner_embedding = self.positional_embedding_layer.call_with_coords(
+            coords, self.input_image_size
+        )
+        top_left_embedding = (
+            corner_embedding[:, 0, :] + self.top_left_corner_embed.weights[0]
+        )
+        bottom_right_embedding = (
+            corner_embedding[:, 1, :]
+            + self.bottom_right_corner_embed.weights[0]
+        )
+        corner_embedding = ops.stack(
+            [top_left_embedding, bottom_right_embedding], axis=1
+        )
+        return corner_embedding
+
+    def __embed_mask(self, mask):
+        mask_embedding = self.mask_downscaler(mask)
+        return mask_embedding
+
+    def call(self, points=None, labels=None, box=None, mask=None):
```

Comment:
Maybe batch together the statments under the `if` and `elif` statements? `if points is not None -> get bs -> do point embedding` for example

Reply:
Removed, this code wasn't XLA compatible. I have rewritten this using `ops.cond` and added a vectorized test for it in `keras_cv/models/segmentation/segment_anything/sam_test.py`.